### PR TITLE
Handle weights that do not sum to 100

### PR DESCRIPTION
**Pull request by maxcabrajac**
**What this PR does / why we need it**:

Since https://github.com/argoproj-labs/rollouts-plugin-trafficrouter-contour/pull/61 this plugin refuses to alter httpProxies whose weights sum to values different then 100. On the other hand, contour doesn't enforce any value for this sum (See [countour's documentation](https://projectcontour.io/docs/v1.6.1/httpproxy/#:~:text=HTTPProxy%20weighting%20follows%20some%20specific%20rules%3A)). This PR aims to handle all weight rules specified on contour's docs.

**Special notes for your reviewer**:

If the initial weight sum is very low, the plugin will lose a lot of precision when redirecting to canaries. (This is most likely the reason why the "equal to 100" check was made in the first place). For example: if both initial weights were set to 1, the only canary-stable ratios achievable would be 0%, 50%  and 100%.
To get around that, this PR boots weights by a factor of 100 whenever the weight sum is bellow 100.

When adding tests I noticed that some of them had duplicate names, so I also renamed them to better fit what they were testing.

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

File: pkg/plugin/plugin.go [hunk 1]
```diff
@@ -171,12 +171,63 @@ func (r *RpcPlugin) updateHTTPProxy(
...
+		route.Services[i].Weight = 1
+	}
+}
+
+func weightSum(route *contourv1.Route) int64 {
+	sum := int64(0)
+
+	for _, s := range route.Services {
+		sum += s.Weight
+	}
```
[the remark this answers is not in the document]
Comment (maxcabrajac) [reply pkg/plugin/plugin.go:207]: Fixed on f4c13fc3478320d4eaae7b8c8c36f57d8e116585

File: pkg/plugin/plugin.go [hunk 2]
```diff
@@ -171,12 +171,63 @@ func (r *RpcPlugin) updateHTTPProxy(
...
+
 func createPatch(httpProxy *contourv1.HTTPProxy, rollout *v1alpha1.Rollout, canaryWeightPercent int32) ([]byte, types.PatchType, error) {
 	oldData, err := json.Marshal(httpProxy.DeepCopy())
 	if err != nil {
 		return nil, "", fmt.Errorf("failed to marshal the current configuration: %w", err)
 	}
 
+	for i := range httpProxy.Spec.Routes {
+		normalizeWeights(&httpProxy.Spec.Routes[i])
+	}
```
[the remark this answers is not in the document]
Comment (maxcabrajac) [reply pkg/plugin/plugin.go]: Applied with a minor tweak on 43718cad76a9ac03b17a6d13ffb9032c99602211

File: pkg/plugin/plugin.go [hunk 3]
```diff
@@ -262,22 +313,9 @@ func getRouteServices(httpProxy *contourv1.HTTPProxy, rollout *v1alpha1.Rollout)
...
-		// the total weight must equals to 100
-		if otherWeight+canarySvc.Weight+stableSvc.Weight != 100 {
-			return nil, nil, nil, fmt.Errorf("the total weight must equals to 100")
-		}
-
 		canarySvcs = append(canarySvcs, canarySvc)
 		stableSvcs = append(stableSvcs, stableSvc)
-		totalWeights = append(totalWeights, 100-otherWeight)
+		totalWeights = append(totalWeights, canarySvc.Weight + stableSvc.Weight)
 	}
```
[the remark this answers is not in the document]
Comment (maxcabrajac) [reply pkg/plugin/plugin.go:356]: if a weight sum of 0 goes past the weight normalization function it is explicitly set by the user (the unset case is handled by `setDefaultWeights`) and should be kept as is. 

Weight normalization was added to verifyWeight on f9f2d395269603ef0522a3a66f82873c5cdd5c9e
